### ISSUE TITLE
CAMonitor is not canceled when closing non-top level OPI

### ISSUE THREAD

**Issue by krisztianloki** [ControlSystemStudio/phoebus#1599]
When an OPI is opened from a top level OPI monitors are created as expected but when this OPI is then later closed (while the top level one is still open) the IOC continues to send value changed events.
Setting the priority of  `EventCancelRequest` in jca seems to fix the problem but I'm not sure that is the only way.

**Comment by kasemir** [comment]
Do you have an example using `loc:` or `sim:` PVs? Do the PVs still show up in Applications, Debug, PV List? Or is the PV fundamentally closed, but some underlying Channel Access subscription remains?

**Comment by krisztianloki** [comment]
The PVs no longer show up in Applications / Debug / PV List but I can see the events coming in using wireshark. I can also see that an Event cancel request is never sent to the IOC.

**Comment by kasemir** [comment]
Do you see the same with Applications, Display, Probe? Or only with a two-level display? In that case, please create a simple, ideally single-PV example.

**Comment by krisztianloki** [comment]
No, with Probe there are no events coming in after the probe is closed.

**Comment by kasemir** [comment]
Well, then please create a display-based example.

As long as the reference-counted PV is closed, as seems to be the case based on the Applications / Debug / PV List, I would expect the underlying JCA PV to also be cleaned up, but maybe there's something different about your two-level display case.

**Comment by kasemir** [comment]
Strictly speaking the subscription/event/monitor is indeed not cancelled.
`JCA_PV` `close()` will simply `dispose` the underlying JCA channel, without specifically `unsubscribe`-ing, because disposing encompasses the cleanup of subscriptions, as can be seen in the `Probe` example.
.. at least that was the idea, so need a reproducible example to look into it deeper.

**Comment by kasemir** [comment]
I've tried running `softIoc -m N='' -d core/pva/src/test/resources/demo.db` to get example PVs like 'ramp' and 'saw', then creating a 'top.bob' which displays one of the PVs, with button to open 'sub.bob' which displays another.
While playing with opening and closing these displays, I'm checking `casr` on the IOC to monitor the number of connections, and the connection is always closed with the last display, no dangling monitors.

**Comment by krisztianloki** [comment]
Yes, the connection is closed when the the top level display is closed. But both of the monitors remain active when only popup.bob is closed.
[infinite-monitor.zip](https://github.com/ControlSystemStudio/phoebus/files/5404015/infinite-monitor.zip)

**Comment by krisztianloki** [comment]
I've just noticed that I left out to what priority I set `EventCancelRequest`... it is `Request.SEND_IMMEDIATELY_PRIORITY`. Otherwise the cancel request was just queued and the channel transport is destroyed right after the cancel request is created so it was never actually sent.

**Comment by kasemir** [comment]
Thanks for the example. I now see that `casr` lists "2 channels" when we only have 1 known channel left.
There's nothing special about `top-level` or `popup`. Both connect to the same CA server. Closing one of them removes its PV, but the IOC still keeps the subscription. Closing both results in the cleanup.

Will look at explicitly calling `unsubscribe`, `flush` and/or the event priority.

**Comment by kasemir** [comment]
How did you change the priority? Build JCA with updated com.cosylab.epics.caj.impl.requests.EventCancelRequest, like this?

```
@Override
public byte getPriority() {
         return Request.SEND_IMMEDIATELY_PRIORITY;
 }
```

Doesn't seem to make a difference, `casr` on the IOC still shows both channels when one of the displays is already closed:
```
epics> casr 2
Channel Access Server V4.13
1 client connected:
    TCP client at 127.0.0.1:56874 '...':
	User '...', V4.13, Priority = 0, 2 Channels
        Channel: 'ramp'
        Channel: 'saw'

```

**Comment by krisztianloki** [comment]
Yes, exactly like that. After that I saw Event cancel in wireshark and the network traffic indicator of my virtual machine no longer showed continuous traffic (my original test case read the voltage of a Beckhoff analog I/O channel)

**Comment by kasemir** [comment]
Well, I'm still unsure about the `casr` output. I'd expect the channel to no longer be listed when the last client disposes the channel, but it remains. The only detail that changes is is the "on eventq" which goes to 0:

```
	User '...', V4.13, Priority = 0, 2 Channels
	Task Id = 0x7f8ebfd13150, Socket FD = 7
	0.88 secs since last send, 6.75 secs since last receive
	Unprocessed request bytes = 0, Undelivered response bytes = 0
	State = up
        Channel: 'ramp'
            field_type=DBF_DOUBLE (8 bytes), dbr_type=DBF_DOUBLE, 1 element, no filters
            # on eventq=0, access=rw
        Channel: 'saw'
            field_type=DBF_DOUBLE (8 bytes), dbr_type=DBF_DOUBLE, 1 element, no filters
            # on eventq=1, access=rw
```

Anyway, I'd say this is a bug in JCA. The `dispose()` call does call down into `destroyAllMonitors`, so there really is no need to explicitly cancel them, and even if you try, it won't make a difference unless you also call `context.flushio()` or maybe wait, which all seem worse than fixing JCA, i.e.  https://github.com/epics-base/jca.git .
Do you want to do the PR, since you found the issue after all?

**Comment by krisztianloki** [comment]
Thanks, I will do a PR in a minute.
I guess the channel is not closed for the same reason as the event cancel was not sent:
```
		if (connectionState == ConnectionState.CONNECTED)
		{
			if (transport != null)
			{
				try
				{
					// NOTE: this does not submit message immediately, waits for flush to be triggered somehow.. hmmm.
					// possible deadlock if sent immediately (since lock on this channel instance is held)
					new ClearChannelRequest(transport, channelID, serverChannelID).submit();
				}
				catch (IOException ioex)
				{
					// TODO remove?
					logger.log(Level.SEVERE, "", ioex);
				}
			}
			
			// TODO transport is closed immedately,
			// so it might happen that ClearChannelRequest is not sent/response received
			disconnect(false);
		}
```

**Comment by kasemir** [comment]
Yes, making `ClearChannelRequest` use `Request.SEND_IMMEDIATELY_PRIORITY` will clear the channel right away, just tried that.
Not sure if that should also be updated in JCA, or if it's a feature, the channel-clear is best queued up to reduce the number of network packets...

**Comment by krisztianloki** [comment]
And that "possible deadlock" comment is a bit worrying too...

**Comment by kasemir** [comment]
Another comment "it might happen that ClearChannelRequest is not sent" describes the exact problem that you found, which is reproducible, and results in erroneous network traffic.
Maybe there is a chance of deadlock, but I've not seen it after opening/closing several larger displays with the updated `EventCancelRequest`.
So my leaning would be to at least merge the PR as you have it now.

**Comment by slominskir** [comment]
I'm glad to see others are working to improve CAJ.    I'm still seeing odd errors with CAJ as epics2web is used more than ever due to COVID.   I definitely don't have a good feeling about CAJ - it is not a strong foundation to build on.  Unfortunately the newer ca doesn't seem like it is much different. (mostly the same underneath).   The "might cause deadlock" comment certainly isn't good.   Two major issues with CAJ:

1. There is too much unnecessary surface area in the API - multiple ways to do the same thing - like CAJContext.destroyChannel() vs CAJChannel.destroy() vs CAJChannel.destroyChannel(), some of which are booby trapped.
1. The threading and locking model seems overly complicated with excess locking and weak guarantees - the timeout to avoid deadlock feels like implementers gave up on trying to get threading right.  Timeout may break deadlock free, but may also leave Context is odd state.

What are the implications of the EventCancelRequest issue?   It appears unnecessary value updates may continue to be delivered to the client wasting resources.  Does it also have implications for deadlock?

**Comment by krisztianloki** [comment]
epics-base/jca#33 might have fixed the deadlock issue... could that be the case?
The implication of this event cancel issue is exactly what you described; unnecessary updates are delivered even after the channel is destroyed on the client side. I don't think it has anything to do with the deadlock.

**Comment by slominskir** [comment]
I think there are multiple deadlock opportunities and we're playing a game of whack-a-mole.  I believe Patch 33 was for a deadlock scenario that could occur before __new ClearChannelRequest()__ could be called.   However, I'm going cross-eyed attempting to re-trace that scenario so maybe I'm mistaken.  It appears there are fistfuls of locks acquired at the edge of the public API, but also a deep nesting of method calls that also occasionally pick up various locks as well - recipe for disaster?

Another comment in CAJChannel.java that does not inspire confidence:
`// TODO consider using WeakHashMap (and call Channel.destroy() in finalize() method)`

**Comment by krisztianloki** [comment]
As far as I can see both `EventCancelResponse` and `ClearChannelResponse` is a no-op so at least those won't try to acquire some lock already held...
What I don't understand is that why the implementation accepts monitor events for channels that are already destroyed? Shouldn't it actively cancel the subscription when an unsolicited event is received?

**Comment by kasemir** [comment]
Well, based on timing, I'd expect this to happen a lot:

Client subscribes
Server sends update
Server sends update
...
Client cancels subscription
Server had just sent a few more updates which now reach the client

So it would be quite common to get a few more monitor events after cancellation.
Makes sense not to print a severe warning, since it's almost expected.
Also makes sense not to cancel again.

**Comment by krisztianloki** [comment]
That is true... but maybe after a "cool down" period it should assume that the previous cancel failed (or wasn't even attempted) and resend it.

**Comment by kasemir** [comment]
Well, we now cancel the subscription, and all's fine.
I agree that it's good to have some robustness in the protocol, but 'cancel' is sent via TCP, so it's not lost.
When events arrive after 'cancel', after the channel is destroyed, I think the most robust reaction is to simply ignore.
Maybe log a warning so we can fix additional issues. But sending additional messages, i.e. creating more network traffic seems to go the wrong way.

**Comment by krisztianloki** [comment]
When do you think a new jca will be released? Should we keep this issue open until then?

**Comment by kasemir** [comment]
I'm interested in a jca release because of https://github.com/epics-base/jca/pull/53, and this event-cancel fix would also be nice. @kunal had suggested "within 2 weeks". Is that still likely, or is the EPICS meeting adding a week?

**Comment by shroffk** [comment]
> "within 2 weeks"

I had taken into account the epics meeting. I hope to perform the release in the beginning of next week.

**Comment by kasemir** [comment]
Good news, thanks!

**Comment by shroffk** [comment]
There is a new release 2.4.6
I have a PR open (epics 7.0.7)... @krisztianloki can you test the branch

**Comment by krisztianloki** [comment]
For some reason I still have `jca-2.4.5.jar` in `phoebus-product/target/lib`... tried deleting it from `.m2/repository` but maven downloaded it again...
`mvn verify -f dependencies/pom.xml` says that `[INFO] jca-2.4.6.jar already exists in destination.` so I'm not sure what is going on...

**Comment by shroffk** [comment]
@krisztianloki are you sure you are testing with the correct branch?
https://github.com/ControlSystemStudio/phoebus/tree/epics_7.0.7

**Comment by krisztianloki** [comment]
Yes:
```
commit 4f396d0ae3068e6e606006787ea4ccc813bb0931 (HEAD, refs/remotes/origin/epics_7.0.7, refs/heads/epics_7.0.7)
Author: Kunal Shroff <kunalshroff9@gmail.com>
Date:   Tue Oct 27 13:09:37 2020 -0400

    Preparing for epics 7.0.7 release

```
and I have `epics-core-7.0.7.pom` in `phoebus-product/target/lib`

**Comment by krisztianloki** [comment]
Took a closer look... in `phoebus-product/target/lib` these seem to be up-to-date:
```
epics-core-7.0.7.pom
epics-ntypes-0.3.6.jar
epics-pvdata-6.1.6.jar
vtype-1.0.4.jar
```
and these don't:
```
epics-pvaccess-5.1.5.jar
jca-2.4.5.jar
```

**Comment by krisztianloki** [comment]
I think I've found the culprit:
```
[INFO] ------------------------------------------------------------------------
[INFO] Building app-save-and-restore-ui 4.6.4-SNAPSHOT
[INFO] ------------------------------------------------------------------------
Downloading: https://oss.sonatype.org/content/repositories/snapshots/org/epics/gpclient-core/1.0.4/gpclient-core-1.0.4.pom
Downloading: https://repo.maven.apache.org/maven2/org/epics/gpclient-core/1.0.4/gpclient-core-1.0.4.pom
Downloaded: https://repo.maven.apache.org/maven2/org/epics/gpclient-core/1.0.4/gpclient-core-1.0.4.pom (817 B at 7.5 kB/s)
Downloading: https://oss.sonatype.org/content/repositories/snapshots/org/epics/gpclient/1.0.4/gpclient-1.0.4.pom
Downloading: https://repo.maven.apache.org/maven2/org/epics/gpclient/1.0.4/gpclient-1.0.4.pom
Downloaded: https://repo.maven.apache.org/maven2/org/epics/gpclient/1.0.4/gpclient-1.0.4.pom (4.0 kB at 34 kB/s)
Downloading: https://oss.sonatype.org/content/repositories/snapshots/org/epics/epics-parent/7.0.6/epics-parent-7.0.6.pom
Downloading: https://repo.maven.apache.org/maven2/org/epics/epics-parent/7.0.6/epics-parent-7.0.6.pom
Downloaded: https://repo.maven.apache.org/maven2/org/epics/epics-parent/7.0.6/epics-parent-7.0.6.pom (17 kB at 152 kB/s)
Downloading: https://oss.sonatype.org/content/repositories/snapshots/org/epics/vtype/1.0.3/vtype-1.0.3.pom
Downloading: https://repo.maven.apache.org/maven2/org/epics/vtype/1.0.3/vtype-1.0.3.pom
Downloaded: https://repo.maven.apache.org/maven2/org/epics/vtype/1.0.3/vtype-1.0.3.pom (575 B at 5.1 kB/s)
Downloading: https://oss.sonatype.org/content/repositories/snapshots/org/epics/epics-vtype-all/1.0.3/epics-vtype-all-1.0.3.pom
Downloading: https://repo.maven.apache.org/maven2/org/epics/epics-vtype-all/1.0.3/epics-vtype-all-1.0.3.pom
Downloaded: https://repo.maven.apache.org/maven2/org/epics/epics-vtype-all/1.0.3/epics-vtype-all-1.0.3.pom (2.8 kB at 27 kB/s)
Downloading: https://oss.sonatype.org/content/repositories/snapshots/org/epics/epics-util/1.0.3/epics-util-1.0.3.pom
Downloading: https://repo.maven.apache.org/maven2/org/epics/epics-util/1.0.3/epics-util-1.0.3.pom
Downloaded: https://repo.maven.apache.org/maven2/org/epics/epics-util/1.0.3/epics-util-1.0.3.pom (3.0 kB at 29 kB/s)
Downloading: https://oss.sonatype.org/content/repositories/snapshots/org/epics/gpclient-ca/1.0.4/gpclient-ca-1.0.4.pom
Downloading: https://repo.maven.apache.org/maven2/org/epics/gpclient-ca/1.0.4/gpclient-ca-1.0.4.pom
Downloaded: https://repo.maven.apache.org/maven2/org/epics/gpclient-ca/1.0.4/gpclient-ca-1.0.4.pom (1.8 kB at 17 kB/s)
Downloading: https://oss.sonatype.org/content/repositories/snapshots/org/epics/jca/2.4.5/jca-2.4.5.pom
Downloading: https://repo.maven.apache.org/maven2/org/epics/jca/2.4.5/jca-2.4.5.pom
Downloaded: https://repo.maven.apache.org/maven2/org/epics/jca/2.4.5/jca-2.4.5.pom (8.9 kB at 80 kB/s)
Downloading: https://oss.sonatype.org/content/repositories/snapshots/org/epics/gpclient-pva/1.0.4/gpclient-pva-1.0.4.pom
Downloading: https://repo.maven.apache.org/maven2/org/epics/gpclient-pva/1.0.4/gpclient-pva-1.0.4.pom
Downloaded: https://repo.maven.apache.org/maven2/org/epics/gpclient-pva/1.0.4/gpclient-pva-1.0.4.pom (1.8 kB at 16 kB/s)
Downloading: https://oss.sonatype.org/content/repositories/snapshots/org/epics/epics-pvaccess/5.1.5/epics-pvaccess-5.1.5.pom
Downloading: https://repo.maven.apache.org/maven2/org/epics/epics-pvaccess/5.1.5/epics-pvaccess-5.1.5.pom
Downloaded: https://repo.maven.apache.org/maven2/org/epics/epics-pvaccess/5.1.5/epics-pvaccess-5.1.5.pom (17 kB at 152 kB/s)
Downloading: https://oss.sonatype.org/content/repositories/snapshots/org/epics/epics-pvdata/6.1.5/epics-pvdata-6.1.5.pom
Downloading: https://repo.maven.apache.org/maven2/org/epics/epics-pvdata/6.1.5/epics-pvdata-6.1.5.pom
```

**Comment by shroffk** [comment]
great...I must have overlooked this.

I updated the target and stopped.
I will update my PR

**Comment by shroffk** [comment]
Yup... that was it.
I did a clean and build and the jca version was correctly updated to 2.4.6

I pushed the fix to the branch.

**Comment by krisztianloki** [comment]
Yes, epics_7.0.7 fixes this issue. Thanks!

**Comment by shroffk** [comment]
Well you fixed it in jca so thanks! :)
We can close this ticket once I finish the merge

**Comment by krisztianloki** [comment]
The fix is merged with #1610
